docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,241 @@
+
+#### Table
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `addprops(key: typelib, value: typelib)` | `nil` | adds a key, value pair to self's table properties; if the same key exists already, overrides the value |
+| `delprops(key: typelib)` | `nil` | removes the key from self's table properties along with the value associated with it; if the key doesn't exist, nothing happens |
+| `getprops()` | `{typelib: typelib}` | returns a table of self's table properties (e.g. `{["age"] = 20}` will return `{typelib.getstringsingleton("age") = typelib.getnumber()}`) |
+| `settableindexer(indexty: typelib, indexresultty: typelib)` | `nil` | sets self's index type to the first argument and index result type to the second |
+| `getindextype()` | `typelib` | returns self's index type |
+| `getindexresulttype()` | `typelib` | returns self's index result type |
+| `setsealed(arg: boolean)` | `nil` | sets the table's sealed status. If true, the table becomes a sealed table. Default initialization of tables is unsealed tables (aka setsealed(false)) |
+| `issealed()` | `boolean` | returns the seal status of self |
+
+#### Metatable
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `settable(arg: typelib)` | `nil` | sets self's table to the argument; the argument needs to be `istable()` |
+| `gettable()` | `typelib` | returns self's runtime representation of the type `table` |
+| `setmetatable(arg: typelib)` | `nil` | sets self's metatable to the argument; the argument needs to be `ismetatable()` |
+| `getmetatable()` | `typelib` | returns self's runtime representation of `Metatable` |
+
+#### Function
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `setargtypes(arg: {typelib} \| typelib)` | `nil` | sets self's argument types to the argument, where an array implies a TypePack and the latter implies a Variadic |
+| `getargtypes()` | `{typelib} \| typelib` | returns the runtime representation of self's argument type |
+| `setreturntypes(arg: {typelib} \| typelib)` | `nil` | sets self's return types to the argument, where an array implies a TypePack and the latter implies a Variadic |
+| `getreturntypes()` | `{typelib} \| typelib` | returns the runtime representation of self's return type |
+
+#### Union
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getoptions()` | `table` | returns an array of types that the self's union can represent. For instance, `string \| number` returns `{typelib.getstring(), typelib.getnumber()}` |
+
+#### Intersection
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getparts()` | `table` | returns an array of types represented by self's intersection. For instance, `string & number` returns `{typelib.getstring(), typelib.getnumber()}` |
```

Comment:
Can we rename them to `components` or `constituents` or something? The names `parts` and `options` are an implementation detail in my opinion.

Reply:
Also, we should offer `__iter` metamethod for them. If you have a union, you absolutely want to iterate over it. No other way around it.